#!/bin/bash

SECONDS=0

###################################################################################### PREPROCESSING ###################################################################################################

# Download reference file

wget http://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/hg38.fa.gz

gunzip hg38.fa.gz

# Index reference file if not already

samtools faidx reference/hg38.fa


# create a sequence dictionary for the reference 

gatk CreateSequenceDictionary -R reference/hg38.fa

# Mark Duplicates in bam files:

gatk MarkDuplicates -I samples_bam/S002.bam -O samples_bam/mark_dup_files/S002_marked_duplicates.bam -M samples_bam/mark_dup_files/S002_marked_dup_metrics.txt

gatk MarkDuplicates -I samples_bam/S033.bam -O samples_bam/mark_dup_files/S033_marked_duplicates.bam -M samples_bam/mark_dup_files/S033_marked_dup_metrics.txt

gatk MarkDuplicates -I samples_bam/S043.bam -O samples_bam/mark_dup_files/S043_marked_duplicates.bam -M samples_bam/mark_dup_files/S043_marked_dup_metrics.txt


# Sort cancer samples BAM file:

gatk SortSam -I samples_bam/mark_dup_files/S002_marked_duplicates.bam -O samples_bam/sorted_bam/S002_sorted.bam -SORT_ORDER coordinate

gatk SortSam -I samples_bam/mark_dup_files/S033_marked_duplicates.bam -O samples_bam/sorted_bam/S033_sorted.bam -SORT_ORDER coordinate

gatk SortSam -I samples_bam/mark_dup_files/S043_marked_duplicates.bam -O samples_bam/sorted_bam/S043_sorted.bam -SORT_ORDER coordinate

# Indexing sorted bam files

samtools index samples_bam/sorted_bam/S002_sorted.bam

samtools index samples_bam/sorted_bam/S033_sorted.bam

samtools index samples_bam/sorted_bam/S043_sorted.bam

###############################################################################		OPTIONAL	 ##########################################################################################


# Base Recalibration (BQSR)

#gatk BaseRecalibrator -I samples_bam/sorted_bam/S002_sorted.bam -R reference/hg38.fa --known-sites known_sites.vcf -O samples_bam/bsqr_out/S002_recal_data.table
#gatk BaseRecalibrator -I samples_bam/sorted_bam/S033_sorted.bam -R reference/hg38.fa --known-sites known_sites.vcf -O samples_bam/bsqr_out/S033_recal_data.table
#gatk BaseRecalibrator -I samples_bam/sorted_bam/S043_sorted.bam -R reference/hg38.fa --known-sites known_sites.vcf -O samples_bam/bsqr_out/S043_recal_data.table

# Apply the recalibration to BAM file:

#gatk ApplyBQSR -R reference/hg38.fa -I samples_bam/sorted_bam/S002_sorted.bam --bqsr-recal-file samples_bam/bsqr_out/S002_recal_data.table -O recalibrated.bam
#gatk ApplyBQSR -R reference/hg38.fa -I samples_bam/sorted_bam/S033_sorted.bam --bqsr-recal-file samples_bam/bsqr_out/S033_recal_data.table -O recalibrated.bam
#gatk ApplyBQSR -R reference/hg38.fa -I samples_bam/sorted_bam/S043_sorted.bam --bqsr-recal-file samples_bam/bsqr_out/S043_recal_data.table -O recalibrated.bam

###############################################################################		OPTIONAL	 ##########################################################################################

# Download the genomad.hg38 for germline mutation filtering 


wget https://www.bcgsc.ca/downloads/morinlab/reference/af-only-gnomad.hg38.vcf.gz

gatk IndexFeatureFile -F af-only-gnomad.hg38.vcf.gz


# Variant Calling on cancer Samples: Run the GATK mutect2 caller on the cancer samples with germline resource gnomad.hg38

#gatk Mutect2 -R reference/hg38.fa -I samples_bam/S002.bam -I samples_bam/S033.bam -I samples_bam/S043.bam -L bedfile/Illumina_Exome_TargetedRegions_v1.2.hg38.bed --germline-resource af-only-gnomad.hg38.vcf.gz -O output/tumor_sample.vcf.gz

gatk Mutect2 -R reference/hg38.fa -I samples_bam/sorted_bam/S002_sorted.bam -L bedfile/Illumina_Exome_TargetedRegions_v1.2.hg38.bed --germline-resource af-only-gnomad.hg38.vcf.gz --f1r2-tar-gz sample_variant2/S002_tumor-f1r2.tar.gz -O sample_variant2/S002_tumor_variants.vcf

gatk Mutect2 -R reference/hg38.fa -I samples_bam/sorted_bam/S033_sorted.bam -L bedfile/Illumina_Exome_TargetedRegions_v1.2.hg38.bed --germline-resource af-only-gnomad.hg38.vcf.gz --f1r2-tar-gz sample_variant2/S033_tumor-f1r2.tar.gz -O sample_variant2/S033_tumor_variants.vcf

gatk Mutect2 -R reference/hg38.fa -I samples_bam/sorted_bam/S043_sorted.bam -L bedfile/Illumina_Exome_TargetedRegions_v1.2.hg38.bed --germline-resource af-only-gnomad.hg38.vcf.gz --f1r2-tar-gz sample_variant2/S043_tumor-f1r2.tar.gz -O sample_variant2/S043_tumor_variants.vcf

#--ob-priors output/tumor_sample-f1r2.tar.gz: This argument specifies a file containing the artifact prior probabilities for each read orientation. The f1r2.tar.gz file is generated by Mutect2 and contains statistics about the orientation bias observed in the sequenced data. FilterMutectCalls uses this information to apply the orientation bias filter, which can help reduce false-positive calls that result from systematic sequencing artifacts related to the orientation of the DNA fragments



# Filtering Germline Variants:

#Mutect2â€™s optional error correction feature, which uses the fragment-one and fragment-two (F1R2) tar file to learn the orientation bias model

#gatk FilterMutectCalls -V sample_variant2/S002_tumor_variants.vcf -R reference/hg38.fa --ob-priors sample_variant2/S002_tumor-f1r2.tar.gz -O output/S002_tumor_variants_filtered.vcf.gz

#Without orientation bias model

gatk FilterMutectCalls -V sample_variant2/S002_tumor_variants.vcf -R reference/hg38.fa  -O sample_variant2/filtered/S002_tumor_variants_filtered.vcf.gz

gatk FilterMutectCalls -V sample_variant2/S033_tumor_variants.vcf -R reference/hg38.fa  -O sample_variant2/filtered/S033_tumor_variants_filtered.vcf.gz

gatk FilterMutectCalls -V sample_variant2/S043_tumor_variants.vcf -R reference/hg38.fa  -O sample_variant2/filtered/S043_tumor_variants_filtered.vcf.gz




#Annotation and Filtering: 
#Funcotator will add annotations to each variant, including gene name, variant classification, protein change, and other relevant details. It does not inherently filter variants like a PoN or af-only-gnomad.vcf.gz would, but it can add information that helps you to make informed decisions about which variants are likely to be somatic mutations with functional impacts.


## CREATING annotated VCF files:

gatk Funcotator --variant sample_variant2/filtered/S002_tumor_variants_filtered.vcf.gz --reference reference/hg38.fa --ref-version hg38 --data-sources-path funcotator_dataSources.v1.8.hg38.20230908s --output funcotator_output/S002_filtered_tumor_variants_annotated.vcf --output-file-format VCF --disable-sequence-dictionary-validation

gatk Funcotator --variant sample_variant2/filtered/S033_tumor_variants_filtered.vcf.gz --reference reference/hg38.fa --ref-version hg38 --data-sources-path funcotator_dataSources.v1.8.hg38.20230908s --output funcotator_output/S033_filtered_tumor_variants_annotated.vcf --output-file-format VCF --disable-sequence-dictionary-validation

gatk Funcotator --variant sample_variant2/filtered/S043_tumor_variants_filtered.vcf.gz --reference reference/hg38.fa --ref-version hg38 --data-sources-path funcotator_dataSources.v1.8.hg38.20230908s --output funcotator_output/S043_filtered_tumor_variants_annotated.vcf --output-file-format VCF --disable-sequence-dictionary-validation



duration=$SECONDS
echo "$(($duration / 60)) minutes and $(($duration % 60)) seconds elapsed."

